Switch linear solvers to iterable actions and remove unnecessary compute tags

nilsvu opened the pull request:
## Proposed changes

This is a housekeeping PR to address #2425, #2186 (at least partially) and reduce the number of compute tags in the linear solver code. These changes reduce the complexity of the code significantly and make it (hopefully) easier to understand. Here's the changes:

- Use `receive_data` and iterable actions for the parallel linear solvers instead of terminating the algorithm and restarting it with simple actions (came up in a review by @moxcodes and more recently by @nilsdeppe, discussed in #2186 and #2425).

  Note that this applies to the element array. Reductions still invoke simple actions on the residual monitor singletons, since synchronizing actions lists and their phases between components is currently quite painful (e.g. we have many components that hard-code the init and registration phase names because of this).
- The change in architecture now allows to loop the parallel linear solver algorithm directly in the iterable actions, instead of relying on `RepeatUntil`. So parallel linear solvers now just expose a single action list that does the solve. The looping actions are nicely indented in the action list because they are template parameters (see e.g. `SolvePoissonProblem.hpp`), which addresses my concern in #2356 that loops should be visible in the PDALs.
- Fix includes (closes #2425) and linking.

The last commit also removes a bunch of unnecessary compute tags from linear solvers, which reduces the complexity of the  code significantly. I could factor that commit into an extra PR if reviewers prefer.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Fix a typo

nilsvu committed:
Remove unnecessary compute tags from linear solvers

The linear solvers used compute tags way too extensively. This commit
replaces most of the compute tags in the linear solvers with simple
function calls. As a result, also some of the simple tags can be
removed. This reduces the complexity of the code quite significantly.

Note that some of the bookkeeping simple tags on the residual monitor
singleton components are replaced by additional reduction data, i.e.
iteration IDs are sent along with the data. These few additional bytes
propably won't affect performance, since the cost is presumably
dominated by the communication overhead, not the amount of data sent.

nilsvu commented:
@moxcodes and @nilsdeppe could you review this PR, since it addresses concerns both of you voiced in earlier reviews?

moxcodes commented:
Sure, I'll review.

nilsvu commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -31,11 +32,43 @@ class TaggedTuple;
+                       const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                       const ArrayIndex& /*array_index*/) noexcept {
+    const auto& inbox = get<InboxTag>(inboxes);
+    return inbox.find(db::get<TemporalIdTag>(box)) != inbox.end();
```
Please just let me know how you would prefer this pattern be handled. The options I can think of are (in order of decreasing code duplication):

1. Copy-paste the `inbox_tags`, `is_ready` function and `extract_from_inbox` function to the 6 places in this PR where they are used.
2. Factor them into `detail` namespaces as is currently done in this PR.
3. Factor them into files in `Parallel/`, as is done in #2489.

nilsdeppe commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -82,52 +115,42 @@ struct PrepareSolve {
 template <typename FieldsTag, typename OptionsGroup>
-struct PrepareStep {
+struct InitializeHasConverged
+    : Receive<Tags::InitialHasConverged<OptionsGroup>,
```
Could you try adding a

```cpp
using is_ready = Receive<..>::is_ready;
```
into the action? That would make it clear what this base class is doing. Also, maybe a clearer name than `Receive` can be thought of :)

nilsvu replied:
Here's the comment from above, since we probably typed at the same time: https://github.com/sxs-collaboration/spectre/pull/2495#discussion_r487209170
> Please just let me know how you would prefer this pattern be handled. The options I can think of are (in order of decreasing code duplication):
> 
> 1. Copy-paste the `inbox_tags`, `is_ready` function and `extract_from_inbox` function to the 6 places in this PR where they are used.
> 2. Factor them into `detail` namespaces as is currently done in this PR.
> 3. Factor them into files in `Parallel/`, as is done in #2489.

nilsvu replied:
I tried the `using is_ready = ...` pattern, but it doesn't work since `is_ready` is a function, not a type.. unless there's some C++ syntax that I missed.

nilsvu replied:
Regarding names all I could think of was `WaitUntilReceived<InboxTag, TemporalIdTag>`

nilsdeppe replied:
```
struct B { template<class T> void f(); };
struct D : B {
      using B::f;      // OK: names a template
//    using B::f<int>; // Error: names a template specialization
      void g() { f<int>(); }
};
```
It should be doable. I've done this before :) It's exactyl how `make_overloader` works.

Why not `SimpleIsReady` or some variant thereof? I'm looking something that indicates it's an `is_ready` supplier

nilsvu replied:
Yep that works, thanks! (no `=`)

`SimpleIsReady` doesn't read as nice, but I like that it makes clear what it is, and I can't think of anything better.

nilsvu replied:
How about `IsReadyWhenReceived<InboxTag, TemporalIdTag>`?

nilsdeppe replied:
That doesn't really narrow down what is checked, though. Literally any `is_ready` struct should be called that then, or am I missing something? I was hoping to have some indication that this is the simplest `is_ready` you can write "It exists". Unsure :shrug:

nilsvu replied:
Agreed. For this PR I inlined the `is_ready` method and the `extract_from_inbox` function for now, so we can move this along. I'm happy to factor them out again either in this PR or in #2489 if you or @moxcodes prefer that, but I'd like to avoid holding up this PR with that discussion.

nilsvu committed:
Add constructor to HasConverged

Used to store the state of an algorithm that has completed
a fixed number of iterations.

nilsvu committed:
Switch linear solvers to iterable actions

- Use `receive_data` and iterable actions for the parallel linear solvers
instead of terminating the algorithm and restarting it with simple actions.
- The change in architecture also allows to loop the parallel linear
solver algorithm directly in the iterable actions, instead of relying on
`RepeatUntil`. So parallel linear solvers now just expose a single
action list that does the solve.
- Fix includes (closes #2425) and linking.

nilsvu committed:
Minor formatting

nilsvu committed:
Allow mock components to add inbox tags for testing

moxcodes commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -82,68 +86,67 @@ struct PrepareSolve {
 
-template <typename FieldsTag, typename OptionsGroup>
+template <typename FieldsTag, typename OptionsGroup, typename Label>
 struct InitializeHasConverged {
```
Sorry, I need to check my understanding of the system here.
So, is it the case that `PrepareSolve` kicks off a reduction that determines the data necessary to compute a global residual, and therefore to determine whether the global solver has converged; Then this action is called to populate the corresponding flag in the individual elements?
Is this communication pattern written down somewhere? I'm a little concerned that hiding away the technical details in the `detail` namespace without careful documentation might make the elliptic code difficult to maintain for developers who aren't familiar with its inner workings.

nilsvu replied:
That's correct. This pattern of reducing to the residual monitor and broadcasting back is documented in `LinearSolver::cg::ConjugateGradient` and `LinearSolver::gmres::Gmres` (it doesn't detail the initial reduction, but explains the reduce-and-broadcast pattern). Do you think additional documentation is needed somewhere?

moxcodes replied:
I see, thanks for pointing me to those. I think I'm happy with that documentation as sufficient for 'user-facing' purposes. I'll have to look a bit more at the linear solver code to understand whether or not I think it's sufficient to help a new maintainer of the elliptic code -- If I find that I'd like to ask for more documentation, I'll file an issue, no need to hold things up here.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/AsynchronousSolvers/ElementActions.hpp:
```diff
@@ -196,11 +204,21 @@ struct PrepareSolve {
-    return {std::move(box)};
+    // Skip steps entirely if the solve has already converged
+    constexpr size_t step_end_index =
+        tmpl::index_of<ActionList, CompleteStep<FieldsTag, OptionsGroup,
```
Would it be too cumbersome to introduce label actions to indicate the beginning and end of steps that are frequently jumped to? My thought is that while it is obvious that I shouldn't add actions after `CompleteStep`, there are other 'final' actions (e.g. `UpdateOperand` in the conjugate gradient) for which that is less obvious, and they're still used for jumping. I think it would be more obvious in the action list that those are points that would be frequently jumped to.

nilsvu replied:
Not sure, I tried to avoid adding extra label actions because the user doesn't see them anyway. The user just adds `Gmres::solve<ApplyOperator>` to their action list. They shouldn't use the actions separately (that's why they are in a detail namespace).

moxcodes replied:
That's reasonable, but here again I'm concerned about maintainability rather than user-facing interface. For instance, if you or someone else needs to insert another action after the one that is currently used as the final action in a step (or before the one used as the start), all of these indices would individually need to be updated. I'd suggest labels, but I won't insist if you still think it doesn't make an improvement.

nilsvu replied:
I see your concern and will keep this in mind. Might add labels in another PR.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ConjugateGradient.hpp:
```diff
@@ -73,107 +77,21 @@ struct ConjugateGradient {
 
   using register_element = tmpl::list<>;
 
-  /*!
```
There's a whole bunch of documentation going away in this commit -- some of things being documented are also going away (or moving somewhere else?), but not all. Should the documentation be moved somewhere instead of just being deleted?

nilsvu replied:
The documentation I deleted was stale or just wrong because it referred to implementation details that changed. I don't think it was ever useful. Instead, I'll probably add a dev guide on how to run linear solves. It could also help to add a protocol that documents the interface that all linear solvers have in common (they have `initialize_element`, `register_element` and `solve` action lists).

moxcodes commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp:
```diff
@@ -232,19 +249,37 @@ struct UpdateOperand {
-                    const ArrayIndex& array_index, const double res_ratio,
-                    const Convergence::HasConverged& has_converged) noexcept {
+  using inbox_tags =
+      tmpl::list<Tags::ResidualRatioAndHasConverged<OptionsGroup>>;
```
I didn't previously realize precisely how much data the linear solves needed to get communicated from reductions -- for my understanding, can you explain why that communication process is necessary? Is it entirely for computing some global residual measures? I understand it's a larger design question that's already been made long ago, it just seems expensive to be doing a lot of broadcasts to every element. Or do I misunderstand what's being done?

nilsvu replied:
The Krylov-subspace solvers (GMRES and CG) work by computing global inner products to build up a basis of the Krylov subspace (that's just K_n(A,b)={b, A b, A^2 b, ..., A^(n-1) b} where A is the linear operator and b is the source in the equations Ax=b that we want to solve). The global inner products are needed to orthogonalize the basis. It's a nice side-effect that the global reductions also yield the residual magnitude, so it can be used to terminate the algorithm when it falls below a tolerance.

On the other hand, the linear solvers I call "asynchronous" don't need global synchronization points to work. For example, the Richardson solver just iterates every element independently, and the Schwarz solver only communicates data on overlap regions with neighboring elements. Therefore, they only dispatch reductions to _observe_ the residual magnitude, but don't wait for broadcasts.

Note that this issue with global synchronization points is one of the reasons why we use a GMRES solver with a (Multigrid-)Schwarz preconditioner for elliptic solves: The outer GMRES solver only runs for ~10 iterations, so the global synchronizations are not a problem, whereas the Schwarz preconditioner that runs in every GMRES step does all the heavy lifting in parallel.

moxcodes replied:
Aha, got it. Thanks for the explanation.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Gmres/ElementActions.hpp:
```diff
@@ -115,19 +118,36 @@ struct NormalizeInitialOperand {
+
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex>
+  static bool is_ready(const db::DataBox<DbTags>& box,
```
I'm beginning to see why you want the inheritance thing from the other PR :P

nilsvu replied:
In this PR let's just repeat this code 6 times. In #2489 we can decide if any refactoring would make it better :)

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/ConjugateGradient/Test_ElementActions.cpp:
```diff
@@ -52,10 +50,13 @@ struct ElementArray {
-                             tmpl::list<LinearSolver::cg::detail::PrepareStep<
-                                 fields_tag, DummyOptionsGroup>>>>;
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Testing,
```
`UpdateFieldValues` has been changed to take a broadcast inbox tag as well -- should that be tested?

nilsvu replied:
I can't currently test actions that do reductions with the action-testing framework (#2310). `UpdateFieldValues` is being tested as part of the algorithm tests though (`Test_ConjugateGradientAlgorithm` for example). But resolving #2308 and #2310 is on my To-Do list.

moxcodes reviewed (commented):
Mostly looking for clarifications on the system design of the elliptic code that I'm less familiar with on this first pass.

nilsvu commented:
Thanks for your review @moxcodes! I replied to your comments. Let me know if anything is still unclear :)

nilsdeppe commented:
@moxcodes please let us know if you're happy with this :)
